src/User/TeamService.php

```diff
@@ -31,8 +46,67 @@ public function countTeams(): int
         return $this->cache['count'];
     }
 
+    public function createNewTeam(string $name): Team
+    {
+        $team = new Team($name);
+        $this->dispatcher->dispatch(new TeamCreateEvent($team));
+
+        return $team;
+    }
+
+    public function saveTeam(Team $team): Team
+    {
+        if(null === $team->getId()) {
+            // invalidate cache only on new teams
+            return $this->saveNewTeam($team);
+        }
+
+        return $this->updateTeam($team);
+    }
+
+    private function saveNewTeam(Team $team): Team
+    {
+        if (null !== $team->getId()) {
+            throw new InvalidArgumentException('Cannot create team, already persisted');
+        }
+
+        $this->validateTeam($team, ['TeamCreate']);
+
+        $this->dispatcher->dispatch(new TeamCreatePreEvent($team));
+        $this->repository->saveTeam($team);
+        $this->dispatcher->dispatch(new TeamCreatePostEvent($team));
+
+        return $team;
+    }
+
     public function hasTeams(): bool
     {
         return $this->countTeams() > 0;
     }
+
+    private function validateTeam(Team $team, array $groups = []): void
+    {
+        $errors = $this->validator->validate($team, null, $groups);
+
+        if ($errors->count() > 0) {
+            throw new ValidationFailedException($errors);
+        }
+    }
+
+    private function updateTeam(Team $team, array $groups = []): Team
```

Comment:
Copy & Paste again 😁 
```suggestion
    private function updateTeam(Team $team): Team
```